quil-rs/src/program/mod.rs

```diff
@@ -379,6 +381,74 @@ impl Program {
+    ///
+    /// The loop is constructed by wrapping the body of the program in classical Quil instructions.
+    /// The given `loop_count_reference` must refer to an INTEGER memory region. The value at the
+    /// reference given will be set to `iterations` and decremented in the loop. The loop will
+    /// terminate when the reference reaches 0. For this reason your program should not itself
+    /// modify the value at the reference unless you intend to modify the remaining number of
+    /// iterations (i.e. to break the loop).
+    ///
+    /// The loop is constructed with [`TargetPlaceholder`]s in order to prevent conflicts
+    /// with existing label names in the program. These placeholders are resolved before
+    /// returning the program.
+    ///
+    /// If `iterations` is 0, then a copy of the program is returned without any changes.
+    pub fn wrap_in_loop(&self, loop_count_reference: MemoryReference, iterations: u32) -> Self {
+        if iterations == 0 {
+            return self.clone();
+        }
+
+        let mut looped_program = self.clone_without_body_instructions();
+        let start_target = TargetPlaceholder::new("__loop-start".to_string());
+        let end_target = TargetPlaceholder::new("__loop-end".to_string());
+
+        looped_program.add_instructions(
+            vec![
+                Instruction::Declaration(Declaration {
+                    name: loop_count_reference.name.clone(),
+                    size: Vector {
+                        data_type: ScalarType::Integer,
+                        length: 1,
+                    },
+                    sharing: None,
+                }),
+                Instruction::Move(Move {
+                    destination: loop_count_reference.clone(),
+                    source: ArithmeticOperand::LiteralInteger(iterations.into()),
+                }),
+                Instruction::Label(Label {
+                    target: Target::Placeholder(start_target.clone()),
+                }),
+            ]
+            .into_iter()
+            .chain(self.body_instructions().cloned())
+            .chain(vec![
+                Instruction::Arithmetic(Arithmetic {
+                    operator: ArithmeticOperator::Subtract,
+                    destination: ArithmeticOperand::MemoryReference(MemoryReference {
+                        name: "__loop-count".to_string(),
+                        index: 0,
+                    }),
+                    source: ArithmeticOperand::LiteralInteger(1),
+                }),
+                Instruction::JumpUnless(JumpUnless {
+                    target: Target::Placeholder(end_target.clone()),
+                    condition: loop_count_reference,
+                }),
+                Instruction::Jump(Jump {
+                    target: Target::Placeholder(start_target),
+                }),
+            ])
+            .collect::<Vec<Instruction>>(),
+        );
+
+        looped_program.resolve_placeholders();
```

Comment:
wise to think of this!

the catch: it will also resolve any existing placeholders existing in the program, which the user may not want. Would it make sense to have the user provide the Label alongside the MemoryReference? That seems like it may be more consistent.

Reply:
Ah yeah, good point. I guess we could use the custom resolver to only resolve the placeholders we created, but I think you make a good point on consistency between how the targets and loop counter are specified. This approach biases towards giving the user control.